### PR TITLE
Windows support for the JNLP connector

### PR DESCRIPTION
**Pull request by gamma**
This PR is for #151. It is a port of the docker-java support for windows JNLP clients.

Currently it only supports cloud provider that are windows docker variants, swarm is not yet supported. The connection works the same way as the Linux variant, so a minimal windows docker image with java should be enough to test it.

Maybe @erik-am can have a look at it?

### CODE REVIEW

**Comment by KostyaSha** [comment]
Remove client argument, introduce in template Windows/Linux switch. Detect OS not from client, but from template variable.

**Comment by gamma** [comment]
I thought about it. Automatic detection is nice. Not sure if another option is for the future. Would be even better if we could check the type of image - even in a swarm it will be distributed according to its operating system - and then decide on the logic.

**Comment by KostyaSha** [comment]
> Automatic detection is nice.

You can't be sure what and how docker will make for swarm and etc.

File: yet-another-docker-plugin/src/main/resources/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher/init.sh [hunk 1]
```diff
@@ -37,7 +37,7 @@ if ! id -u "$JENKINS_USER"; then
     elif [ -x "$(command -v adduser)" ]; then
         adduser -D -h "${JENKINS_HOME}" "$JENKINS_USER"
     else
-        echo "Error: can't add user. useradd or assuser didn't found."
```
Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/resources/com/github/kostyasha/yad/launcher/DockerComputerJNLPLauncher/init.sh]: :DDDDD lol

**Comment by KostyaSha** [comment]
ok, let's keep autodetection. Config option may be added later any time.

**Comment by erik-am** [comment]
It doesn't work yet, but that may be something wrong on my end... Should this plugin work if the original docker-plugin is also installed as a Cloud Provider or would those two clash? I've added both docker-plugin as yet-another-docker-plugin as cloud providers, but for different hosts and different images.

**Comment by KostyaSha** [comment]
Do you have any errors in system log?

**Comment by erik-am** [comment]
I did have errors, but not any error that contain the namespace of this plugin.

I've just given it another go on a fresh Jenkins installation without any plugins, and there it does launch a container on Windows! The JNLP connection didn't work yet, but that's related to the networking of my test environment, so I know why.

I also did not get the 'Docker Folder Certificate' option on my Jenkins machine where I had the other plugin installed simultaneously, and now I do see that option. So I really think the two plugins were conflicting on my main Jenkins machine.

Tomorrow I'll do another test on my main Jenkins machine. I will completely remove `jenkins-plugin`, then install this one, and then see what happens. Will give an update tomorrow! I'm hopeful. 😄

**Comment by KostyaSha** [comment]
YAD doesn't use folder credentials because jenkins as usual handle credentials from credentials. I implemented right credentials handling long time ago.

**Comment by erik-am** [comment]
I have this strange issue that prevents from testing it on my production environment.

On my test server (clean Jenkins install), I can select "Docker Host Certificate Authentication":
test-server

On my production server (which runs the same Jenkins, latest version), I don't get the "Docker Host Certificate Authentication"
prod-server

@KostyaSha Do you have any idea what the issue could be?

**Comment by erik-am** [comment]
I found the solution to my credentials issue. Apparently I had whitelisted my credentials providers. My bad!

@gamma @KostyaSha Everything seems to work fine on my production machine! I configured YAD with two clouds: one Windows host and one Linux host and images for both platforms. It works fine!
Thank you very much for your efforts. 😄  👍

**Comment by KostyaSha** [comment]
Could you squash commits?

**Comment by gamma** [comment]
👍will do tomorrow

**Comment by KostyaSha** [comment]
solved via #163

**Pull request closed without merging**